subsystem/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+        });
+
+        Subsystem::new(action_tx)
+    }
+
+    /// Start a passive subsystem.
+    ///
+    /// A passive subsystem does not interact with the environment on its own. It only serves calls
+    /// from other subsystems.
+    pub fn start_passive<T: 'static + Send>(self, mut obj: T) -> Subsystem<T> {
+        self.start(|mut call_rq, mut shutdown_rq| async move {
+            loop {
+                tokio::select! {
+                    () = shutdown_rq.recv() => { break; }
+                    call = call_rq.recv() => { call(&mut obj); }
+                }
+            }
+        })
+    }
+}
+
+/// Internal action type sent in the channel.
+type Action<T> = Box<dyn FnOnce(&mut T) + Send>;
+
+/// Call request
+pub struct CallRequest<T>(mpsc::Receiver<Action<T>>);
+
+impl<T: 'static + Send> CallRequest<T> {
+    /// Receive an external call to this subsystem.
+    pub async fn recv(&mut self) -> Action<T> {
+        match self.0.recv().await {
+            // We have a call, return it
+            Some(action) => action,
+            // All handles to this subsystem dropped, suspend call handling.
+            None => std::future::pending().await,
+        }
+    }
+}
+
+/// Shutdown request
+pub struct ShutdownRequest(broadcast::Receiver<()>);
+
+impl ShutdownRequest {
+    /// Receive a shutdown request.
+    pub async fn recv(&mut self) {
+        match self.0.recv().await {
+            Err(broadcast::error::RecvError::Lagged(_)) => {
+                panic!("Multiple shutdown broadcast requests issued")
+            }
+            Err(broadcast::error::RecvError::Closed) => {
+                panic!("Shutdown channel sender closed prematurely")
+            }
+            Ok(()) => (),
+        }
+    }
+}
+
+/// Subsystem handle.
+///
+/// This allows the user to interact with the subsystem from the outside. Currently, it only
+/// supports calling functions on the subsystem.
+pub struct Subsystem<T> {
+    // Send the subsystem stuff to do.
+    action_tx: mpsc::Sender<Action<T>>,
+}
+
+impl<T> Clone for Subsystem<T> {
+    fn clone(&self) -> Self {
+        Self {
+            action_tx: self.action_tx.clone(),
+        }
+    }
+}
+
+impl<T: Send + 'static> Subsystem<T> {
+    /// Crate a new subsystem handle.
+    fn new(action_tx: mpsc::Sender<Action<T>>) -> Self {
+        Self { action_tx }
+    }
+
+    /// Dispatch a function call to the subsystem
+    pub async fn call_mut<R: Send + 'static>(
+        &self,
+        func: impl FnOnce(&mut T) -> R + Send + 'static,
+    ) -> R {
+        let (rtx, rrx) = oneshot::channel::<R>();
+
+        self.action_tx
+            .send(Box::new(move |subsys| {
+                let result = func(subsys);
+                rtx.send(result).ok().expect("Value return channel closed");
+            }))
+            .await
+            .ok()
+            .expect("Target subsystem down");
+
+        rrx.await.expect("Target subsystem down")
```

Comment:
You mentioned in the call that you were working on nested calls but I was wondering can this deadlock the blockchain? I don't know if it can realistically happen with how a blockchain generally works but if subsystem calls can themselves execute subsystem calls, is it possible that A->B->C->A call chain deadlocks the program? So subsystem A calls B which, before returning a value to A, calls C which in turn, before returning a value to B, calls A. Or maybe A calls B and as B is calling C, C independently calls A (as they are scheduled to execute in parallel).

Can any of this actually happen or is this just unwarranted paranoia?

Reply:
I think if we could somehow rule out cyclic calls entirely that would be very nice so I would prefer option 1. I don't know if that is possible as we have two input sources running in parallel (RPC and P2P) so them independently interacting with the blockchain can create a cycle, maybe? Perhaps we could prove with TLA+ that it doesn't happen but I fear we don't have the time :smiley: 

I think the problem with 3 is that it can make refactoring a really scary process because even a hint of carelessness you can create a deadlock.